Marstek integration

MarstekEnergy opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
Add support for Marstek energy storage systems. This integration enables Home Assistant to monitor and control Marstek battery and solar panel systems via local UDP communication.

**Supported devices:**
- Venus A, Venus D, Venus E 3.0 (latest firmware)
- Other Marstek devices with OPEN API support

**Key features:**
- **Automatic device discovery**: UDP broadcast discovery on local network
- **Battery monitoring**: State of charge (SOC), power, voltage, current, temperature
- **Energy storage monitoring**: Real-time status updates via UDP polling
- **Solar panel monitoring**: 4 PV channels with voltage, current, and power data
- **Charge/discharge control**: Three operation modes via services (charge, discharge, stop)
- **Device automation actions**: Integration with Home Assistant automations
- **Local communication**: UDP bidirectional communication on port 30000 (no cloud dependency)

**Communication protocol:**
- UDP protocol on port 30000
- 10-second timeout
- Exponential backoff retry mechanism for mode changes
- Real-time push mechanism via UDP polling

**Commands implemented:**
- Device discovery: `Marstek.GetDevice`
- Battery status: `Bat.GetStatus`
- Energy storage status: `ES.GetStatus`
- Mode setting: `ES.SetMode`
- PV status: `PV.GetStatus` (device-dependent)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
-    Link to documentation pull request: home-assistant/home-assistant.io#41662
- Link to frontend pull request: 

**Requirements:**
- Marstek device must have OPEN API enabled
- Device and Home Assistant must be on the same local network
- UDP port 30000 must be accessible

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

MarstekEnergy committed:
Add Marstek energy storage integration

MarstekEnergy committed:
Replace Chinese comments and documentation with English

pantherale0 commented:
You should remove the device actions and services from this PR and add those in a follow up (see the first rule of https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr)

joostlek reviewed (changes requested):
Hello 👋🏻,

Thanks for bringing this integration to core. A few things that needs to be done for this PR to be reviewable:
1. Please have a proper development environment. Our developemnt environment has tooling to make sure your PR is complete, which it isn't right now.
2. Please make sure that device specific things are stored in a library hosted on PyPI.
3. Let's only add the sensor.py and keep the device_action (which I think shouldn't belong here at all tbf), services and options flow out.
4. You also need a quality_scale.yaml file and reach at least the bronze rule
5. The config flow requires at least 100% unit tests

Thanks!

MarstekEnergy committed:
Add Marstek integration tests and coordinator refactor

MarstekEnergy commented:
@joostlek

jaapp commented:
> Hi @MarstekEnergy , i'm testing this integration, but when the device is intergrated in HA, we see timeouts, but more important, there is an instability connection to the CT003 , after a few minutes, the connection is lost, it takes a while untill the CT connection is back, but then again it looses connection..., disabling this integration, and then the CT is stable again
> 
> I also notice timeouts:
> 
> ```
> 2025-11-15 13:23:39.190 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> 2025-11-15 13:24:03.988 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> 2025-11-15 13:25:11.049 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> 2025-11-15 13:25:38.762 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> ```
> 
> Tested on Venus v2 models, v155.216
> 
> We also see the same behaviour when polling with command line, so its not retaled to the integration, but probably firmware related

Can confirm the lost CT003 connection is probably a Marstek Venus E or CT003 firmware related issues.
Using the Local API appears to impede CT readings and thus the batteries functionality (self consumption mode / AI mode)

We've been testing the Local API using my integration ( https://github.com/jaapp/ha-marstek-local-api ) for a couple of weeks and the issue page is full of firmware related issues:
https://github.com/jaapp/ha-marstek-local-api/issues/15
https://github.com/jaapp/ha-marstek-local-api/issues/19
https://github.com/jaapp/ha-marstek-local-api/issues/20

pergolafabio commented:
> > > > Hi @MarstekEnergy , i'm testing this integration, but when the device is intergrated in HA, we see timeouts, but more important, there is an instability connection to the CT003 , after a few minutes, the connection is lost, it takes a while untill the CT connection is back, but then again it looses connection..., disabling this integration, and then the CT is stable again
> > > > I also notice timeouts:
> > > > ```
> > > > 2025-11-15 13:23:39.190 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> > > > 2025-11-15 13:24:03.988 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> > > > 2025-11-15 13:25:11.049 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> > > > 2025-11-15 13:25:38.762 WARNING (MainThread) [pymarstek.udp] Request timeout: 192.168.0.48:30000
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Tested on Venus v2 models, v155.216
> > > > We also see the same behaviour when polling with command line, so its not retaled to the integration, but probably firmware related
> > > 
> > > 
> > > Hi @pergolafabio, thank you for testing the integration and providing detailed feedback. We understand the timeout and CT003 connection issues you're experiencing. To better understand the current situation, I have a couple of questions:
> > > 
> > > 1. Is the integration able to successfully discover and add your Marstek device? Can you see the device in Home Assistant after the discovery process?
> > > 2. Are you able to get real-time device status updates when the device does respond? Or are the sensors showing as unavailable?
> > >    Regarding the CT003 connection instability, we'll test this issue on our side as well.
> > > 
> > > Thank you for your help!
> > 
> > 
> > Hi, yes, the integration is able to discover the machines I see indeed real time status, but its unstable, the battery doesnt always respond, even polling 60 sec is unstable also
> > thnx
> 
> Hi, I tested with both 10s and 60s as the polling frequency, and found that the request timeout occurrences are not fixed - sometimes it happens every few tens of seconds, sometimes every few minutes. This may be related to the firmware module.
> 
> I also tested with Venus connected to CT003, and haven't found any CT003 connection instability or disconnection issues yet. I will continue testing.
> 
> Thank you for your reply.

ok, we can easily reproduce it, a lots of users are experiencing the same issue, we also see in the marstek app, the CT connection changing from green to gray, it comes back after a while

pergolafabio commented:
@Roos-AID , please dont ask for support here, this is related to the API integration only

pantherale0 commented on homeassistant/components/marstek/__init__.py:
```diff
@@ -0,0 +1,84 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["sensor"]
```
Use the platforms const here:

https://github.com/home-assistant/core/blob/6a8106c0ebc688e6e3cc01401d626fda9b1a7471/homeassistant/components/abode/__init__.py#L48

pantherale0 commented on homeassistant/components/marstek/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    """Set up Marstek from a config entry."""
+    _LOGGER.info("Setting up Marstek config entry: %s", entry.title)
+
+    store = hass.data.setdefault(DOMAIN, {})
```
You only need to use `entry.runtime_data`, there is no need to store anything in `hass.data` for this.

pantherale0 commented on homeassistant/components/marstek/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok and not hass.config_entries.async_entries(DOMAIN):
+        client = hass.data.get(DOMAIN, {}).pop("udp_client", None)
```
A follow up from above, you can remove this if you just use `entry.runtime_data`

Instead for your cleanup, you could just call `entry.runtime_data.async_cleanup()`

pantherale0 commented on homeassistant/components/marstek/coordinator.py:
This seems to be doing quite a bit? Unpacking and parsing responses which I think should be handled by the library.

MarstekEnergy committed:
Merge remote-tracking branch 'upstream/dev' into marstek-integration

MarstekEnergy committed:
Address PR review feedback

- Move PLATFORMS to const.py using Platform enum
- Use entry.runtime_data instead of hass.data
- Remove MarstekRuntimeData dataclass, use MarstekUDPClient directly
- Use _async_abort_entries_match for duplicate detection
- Remove empty MarstekOptionsFlow
- Remove unused async_step_zeroconf
- Move data parsing logic to py-marstek library
- Remove logo.png file
- Update quality_scale.yaml with discovery exempt status
- Remove version from config entry title
- Update dependency to py-marstek==0.2.0
- Fix import statements in test files

MarstekEnergy committed:
Address Copilot AI review feedback

- Remove unused command constants (moved to pymarstek library)
- Fix entity name formatting to use sentence case
- Remove redundant debug logging
- Fix string case issues in strings.json
- Remove unused STEP_USER_DATA_SCHEMA
- Add DIAGNOSTIC category to device info sensors
- Add translation_key support for battery sensor
- Clean up comments in const.py
- Remove README.md (should be in home-assistant.io repo)
- Fix EntityCategory import path

MarstekEnergy committed:
Merge remote-tracking branch 'upstream/dev' into marstek-integration

MarstekEnergy commented:
@frenck  Hi,  I have addressed all the Copilot AI feedback. The code is ready for review.

pergolafabio commented:
> > ok, we can easily reproduce it, a lots of users are experiencing the same issue, we also see in the marstek app, the CT connection changing from green to gray, it comes back after a while
> 
> Today, I used the integrated control Venus E v2 device and connected to CT003. I was able to reproduce this issue. The CT would automatically disconnect, but Venus A did not detect this problem. I will report this to the internal team to attempt to fix this bug.

Hi  @MarstekEnergy , there is now v156 for Venus v1/v2 , does  that include the fix for the CT instability? i will request this trough the app?

MarstekEnergy commented:
> > > > > > > ok, we can easily reproduce it, a lots of users are experiencing the same issue, we also see in the marstek app, the CT connection changing from green to gray, it comes back after a while
> > > > > > 
> > > > > > 
> > > > > > Today, I used the integrated control Venus E v2 device and connected to CT003. I was able to reproduce this issue. The CT would automatically disconnect, but Venus A did not detect this problem. I will report this to the internal team to attempt to fix this bug.
> > > > > 
> > > > > 
> > > > > Hi @MarstekEnergy , there is now v156 for Venus v1/v2 , does that include the fix for the CT instability? i will request this trough the app?
> > > > 
> > > > 
> > > > As @pantherale0 already requested: Please can you move this discussion elsewhere as it doesn't seem relevant to the review of the PR?
> > > 
> > > 
> > > No, this is relevant, its needed in order to test the API and the PR
> > 
> > 
> > This PR is only the integration between HA and Marstek devices. The CT instability issues are firmware-related issues and therefore it has nothing to do with this PR.
> 
> ok, sorry, created an issue to discuss this topic @MarstekEnergy can you provide feedback there?
> 
> #157353

@pergolafabio  Sure, let's discuss this in this issue.

beardhatcode commented on homeassistant/components/marstek/sensor.py:
```diff
@@ -0,0 +1,307 @@
+        MarstekPVSensor(coordinator, device_info, pv_channel, metric_type)
+        for pv_channel in range(1, 5)
+        for metric_type in ("power", "voltage", "current", "state")
+    )
```
Not all devices support this right?


```suggestion
    if device_info.get("device") in [ ... ]
        sensors.extend(
            MarstekPVSensor(coordinator, device_info, pv_channel, metric_type)
            for pv_channel in range(1, 5)
            for metric_type in ("power", "voltage", "current", "state")
        )
```

nickles-lee commented on homeassistant/components/marstek/sensor.py:
```diff
@@ -0,0 +1,307 @@
+
+    device_info = {
+        "ip": config_entry.data["host"],
+        "mac": config_entry.data.get("mac", ""),
```
I've observed two batteries connected to the same wifi AP returning the same MAC value, which _isn't_ a problem with the BLE-MAC. 

It might be good to standardize on that as the device-specific identifier, also given that this is what Marstek uses to uniquely identify the batteries in the app & for warranty/serial number purposes.

MarstekEnergy committed:
Merge remote-tracking branch 'upstream/dev' into marstek-integration

MarstekEnergy committed:
Address PR review feedback: use DeviceInfo class and move attributes to class level

- Replace dict with DeviceInfo class for device_info
- Move MarstekBatterySensor attributes to class level for consistency

MarstekEnergy committed:
Merge remote-tracking branch 'upstream/dev' into marstek-integration

MarstekEnergy committed:
Address PR review feedback: use BLE-MAC as device identifier, fetch device data, add DHCP discovery

- Use BLE-MAC (or MAC/WiFi-MAC) as device identifier instead of IP for stability (beardhatcode & mik-laj feedback)
- Fetch device information from device itself, not from config_entry (mik-laj feedback)
- Create coordinator in __init__.py instead of sensor.py (mik-laj feedback)
- Add DHCP discovery support to handle IP address changes automatically (mik-laj feedback)
- Move sensor attributes to class level for consistency (mik-laj feedback)
- Add automatic IP change detection and device rediscovery in coordinator
- Update device and entity names when IP changes
- Remove IP address from sensor names for cleaner display

MarstekEnergy commented on homeassistant/components/marstek/sensor.py:
```diff
@@ -0,0 +1,308 @@
+
+    device_info = {
+        "ip": config_entry.data["host"],
+        "mac": config_entry.data.get("mac", ""),
```
Thanks for the suggestion! I've implemented fetching device information from the device itself during the `async_setup_entry` process. The integration now retrieves the latest device information directly from the device, ensuring data freshness and automatic IP change detection.

MarstekEnergy committed:
Fix startup hang: optimize DHCP config and discovery logic

- Remove broad macaddress matcher from DHCP config to prevent startup hang
- Use hostname and registered_devices for more efficient DHCP matching
- Optimize device discovery: use cache first, add timeout protection
- Prevent blocking startup by prioritizing cached discovery

mik-laj commented on homeassistant/components/marstek/__init__.py:
```diff
@@ -0,0 +1,214 @@
+            stored_ip,
+            current_ip,
+        )
+        hass.config_entries.async_update_entry(
```
Configuration updates should be done via config flow.

MarstekEnergy committed:
Address PR review: optimize discovery logic and fix linting issues

- Optimize discovery to avoid duplicate broadcasts
- Add automatic cache clearing when wrong MAC detected
- Add comment explaining direct config update in coordinator (UX trade-off)
- Improve error handling with proper exception types
- Update request intervals: ES/PV requests 5s, coordinator polling 10s
- Fix linting issues: remove periods from log messages, fix type hints, fix ruff TRY300/TRY301

mik-laj commented on homeassistant/components/marstek/__init__.py:
```diff
@@ -0,0 +1,294 @@
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+type MarstekConfigEntry = ConfigEntry[
+    tuple[MarstekUDPClient, MarstekDataUpdateCoordinator, dict[str, Any]]
```
Can you create a dataclass for these values? This way we will know what these fields mean.
https://github.com/home-assistant/core/blob/c9351a022ea28546afface55f2bc7d9d0d8de7d7/homeassistant/components/accuweather/coordinator.py#L39-L45

MarstekEnergy commented on homeassistant/components/marstek/__init__.py:
```diff
@@ -0,0 +1,294 @@
+        }
+    except (TimeoutError, OSError, ValueError) as ex:
+        # Connection failed - device IP may have changed
+        # Try discovery to find device by MAC address
```
I've created a MarstekRuntimeData dataclass to store udp_client, coordinator, and device_info. I've refactored the code as follows:
 - Created a new scanner.py file with MarstekScanner class
 - Scanner is initialized in async_setup() (only once, regardless of number of config entries)
 - Removed all discovery code from async_setup_entry() - it now only connects using the configured IP address
- Removed discovery logic from coordinator - it only handles data polling
-Scanner detects IP changes and updates config entries via async_update_entry(), then triggers reload

mik-laj commented on homeassistant/components/marstek/scanner.py:
```diff
@@ -0,0 +1,200 @@
+                        new_ip,
+                    )
+                    # Update config entry via config flow (mik-laj feedback)
+                    self._hass.config_entries.async_update_entry(
```
We should trigger discovery flow. Here is related code in Yeelight.
https://github.com/home-assistant/core/blob/319d6711c4fa4c6f93c4149d2824ad273f6469d8/homeassistant/components/yeelight/scanner.py#L171-L181
And this flow should update config entries.
https://github.com/home-assistant/core/blob/319d6711c4fa4c6f93c4149d2824ad273f6469d8/homeassistant/components/yeelight/config_flow.py#L96-L123

MarstekEnergy committed:
Refactor scanner to use discovery flow for IP updates

- Replace direct config entry updates with discovery_flow.async_create_flow()
- Add async_step_integration_discovery() to handle scanner-triggered discovery
- Implement _async_handle_discovery_with_unique_id() following Yeelight pattern
- Ensures IP updates are handled through config flow as per review feedback

NoRi2909 commented on homeassistant/components/marstek/strings.json:
```diff
@@ -0,0 +1,32 @@
+{
+  "config": {
+    "abort": {
+      "already_configured": "Device is already configured"
```
We have a common string for that:

```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```

MarstekEnergy committed:
Use common string key for already_configured abort reason

pergolafabio commented:
> > ok, we can easily reproduce it, a lots of users are experiencing the same issue, we also see in the marstek app, the CT connection changing from green to gray, it comes back after a while
> 
> Today, I used the integrated control Venus E v2 device and connected to CT003. I was able to reproduce this issue. The CT would automatically disconnect, but Venus A did not detect this problem. I will report this to the internal team to attempt to fix this bug.

Hi @MarstekEnergy , is there an update on this issue with Venus models? I tested already v156, but issue is still there... This will probably merged in 2026.2 , I hope the issue can be fixed before it goes live? Otherwise it will create a lots of issues on GitHub... Since the Venus is a popular model...

MarstekEnergy committed:
Merge remote-tracking branch 'upstream/dev' into marstek-integration

MarstekEnergy committed:
Merge remote-tracking branch 'upstream/dev' into marstek-integration

MarstekEnergy committed:
Update marstek integration

MarckNL commented:
Did you say - No control entities - ?
Help?
The main reason for having the Marstek in HA is to gain control over its
functions, I hope?

Good luck with the progress. I'm eager to try this out on Venus E v3.

Marck Koehler

Op di 13 jan 2026, 17:47 schreef Jojo-A ***@***.***>:

> *Jojo-A* left a comment (home-assistant/core#156012)
> <https://github.com/home-assistant/core/pull/156012#issuecomment-3745354869>
>
> Hi,
> I was just able to install the integration as a custom integration. I have
> a Venus A here.
> Basically that worked quite fine. But some things should be improved in my
> opinion:
>
>    - power of PV 1 seems to be off by the factor of 10
>    - naming pattern for the device (and so the entities) is not good.
>    Currently it is "Marstek Model Firmwareversion IP address". This makes the
>    name unnecessary long and of firmware gets updated, the name is wrong and
>    ALL entities names are wrong. I would suggest to just make it "Marstek
>    Model" or maybe "Marstek Model last-four-digits-of-the-bluetooth-mac"
>    - energy counters are missing for both import, export and solar
>    energy. Of course they can be created by helpers, but I would prefer them
>    to be there ootb.
>    - no control entities. We should have the possibility to set mode and
>    other things like DOD...
>
> Let me know what you think!
>
> Cheers and thank you very much for your effort @MarstekEnergy
> <https://github.com/MarstekEnergy> !
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/156012#issuecomment-3745354869>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BUNZCGSZRVTZIH4FUVWQT3L4GUORLAVCNFSM6AAAAACLM3BO5GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTONBVGM2TIOBWHE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

Jojo-A commented:
> Did you say - No control entities - ?

Yep, none. This is true at least for the Venus A. Maybe you should give it a try for your system as well to verify my observations.

@MarstekEnergy 
Maybe you should have released this integration as a custom integration for beta testing before releasing it to the core.
We highly appreciate your effort and the fact, that Marstek provides an open API and is willing to provide an own integration. That's why many of us (customers) would really love to be involved in the testing process, so that we can provide feedback.

pantherale0 commented:
@Jojo-A @MarckNL - this PR only implements sensors. Core integrations can only implement one platform per PR which is inline with development guidelines.

If reviewing, please keep the reviews to the code that has been proposed in the PR. Feature requests can be created as a discussion in the appropriate category if required once the integration reaches core.

Nontheless, feedback around device / entity naming is very useful, @Jojo-A I agree with you here, the device names should not contain this much information (although technically you can change the device names at the end of the config flow providing the device is correctly picked up).

Jojo-A commented:
> there is a custom on marstek repository

Where? Do you mean the files under https://github.com/MarstekEnergy/home-assistant-core/tree/marstek-integration/homeassistant/components/marstek and placing them under custom-components/marstek?
If yes: this is what I did. But this worked for me only since the most recent commits from yesterday.
I am totally fine with that approach!

My only "complain" is that we should have had the opportunity to install and test this integration <ins>before</ins> creating a PR to the core. Because in my opinion you reach a different groups of users: on the one hand **custom_component** for those who are willing to tinker around and who are aware that things might not work ootb ("testers" and people who are willing to provide feedback) and on the other hand **core integration** for ALL users who (questionably) expect things to work.

Anyway, I didn't want to litter up this conversation :sweat_smile: .

I am so looking forward to see this integration reaching the core! Thank you @MarstekEnergy

slanckma commented:
Hi,

I installed marstek-dev in my HA and have the Venus E 3.0. It loads up fine, but I do have some questions regarding the sensors that it shows:
<img width="1294" height="1212" alt="Screenshot 2026-01-14 at 10 13 14" src="https://github.com/user-attachments/assets/07df5c55-4846-483b-85b1-962c39978976" />

1)Device mode = Auto (what is Auto? This should reflect what is in the Marstek App, no? Which is on "Self Consumption")
2)These PV1-PV4 sensors are not clear to me. What are they referencing? I was expecting Phases 1 to 3 to show up.


Thanks,
Sven.

MarstekEnergy commented:
> > Thank you for the efford, @MarstekEnergy! I am looking forward to use this official integration with my two Venus E 3.0's! 😄
> 
> can't wait ...please update ;-)

@Jojo-A @MarckNL @soepketel-dev 
Thank you for your support!

We’ve observed a specific issue: VenusE 2.0 will disconnect from the CT whenever UDP broadcast is used to send requests. We’re currently waiting for the firmware team to roll out a fix for this problem.

To make sure the integration stays stable, we’re holding off on merging this PR for the time being.

If you need to use the integration right now, feel free to check out this repo: [ha_marstek](https://github.com/MarstekEnergy/ha_marstek)

MarstekEnergy commented:
> Hi @MarstekEnergy , is there an update on this issue with Venus models? I tested already v156, but issue is still there... This will probably merged in 2026.2 , I hope the issue can be fixed before it goes live? Otherwise it will create a lots of issues on GitHub... Since the Venus is a popular model...

Hi, the firmware team is still working on ironing out the VE2 issues. For the time being, I’ll be converting this PR to a draft.
Also, I’ve put together a separate repo tailored for users of other devices: [ha_marstek](https://github.com/MarstekEnergy/ha_marstek).
Once the VE2 problems are sorted out, I’ll circle back and request to merge this PR. Thanks for your patience!

MarstekEnergy commented:
> Hi, I was just able to install the integration as a custom integration. I have a Venus A here. Basically that worked quite fine. But some things should be improved in my opinion:
> 
> * power of PV 1 seems to be off by the factor of 10
> * naming pattern for the device (and so the entities) is not good. Currently it is "Marstek Model Firmwareversion IP address". This makes the name unnecessary long and of firmware gets updated, the name is wrong and ALL entities names are wrong. I would suggest to just make it "Marstek Model" or maybe "Marstek Model last-four-digits-of-the-bluetooth-mac"
> * energy counters are missing for both import, export and solar energy. Of course they can be created by helpers, but I would prefer them to be there ootb.
> * ~no control entities. We should have the possibility to set mode and other things like DOD...~
> 
> Let me know what you think!
> 
> Cheers and thank you very much for your effort @MarstekEnergy !

Hi, Thanks for the feedback! We’ll optimize as suggested, including verifying the PV 1 value for Venus A devices. Check the updated repo here: https://github.com/MarstekEnergy/ha_marstek

MarstekEnergy commented:
> Hi. That's good news! Can you post here the relevant firmware/communication version we need once fixed?
> That way we can ask support team to roll out this specific version ..

ok, as soon as we push out the full release of the fix version for this problem, I’ll post an update in this thread.

Jojo-A commented:
@MarstekEnergy and @pantherale0 
Is this here now also the place to discuss bugs/improvements? Or shall that be moved to the repo here  https://github.com/MarstekEnergy/ha_marstek ?
I just ask to not unnecessarily clutter up this PR...

gizahNL commented:
What's the timeline on this?

LeslytH commented:
Hello,

Myself and I think a lot of others are anxiously waiting till this integration is added into Home Assistant! 

What is the ETA?

KGS501 commented:
For all, who are still waiting for the official integration, I can recommend this custom_integration for now:

https://github.com/ViperRNMC/marstek_venus_modbus

With it you just need to connect you Marstek via Ethernet and can reliably control the device fully locally.
